chain/client/src/chunk_inclusion_tracker.rs

```diff
@@ -102,4 +149,69 @@ impl ChunkInclusionTracker {
         }
         banned_chunk_producers.into_iter().collect_vec()
     }
+
+    fn get_chunk_info(&self, chunk_hash: &ChunkHash) -> Result<&ChunkInfo, Error> {
+        // It should never happen that we are missing the key in chunk_hash_to_chunk_info
+        self.chunk_hash_to_chunk_info
+            .get(chunk_hash)
+            .ok_or(Error::Other(format!("missing key {:?} in ChunkInclusionTracker", chunk_hash)))
+    }
+
+    pub fn chunk_header(&self, chunk_hash: &ChunkHash) -> Result<&ShardChunkHeader, Error> {
+        Ok(&self.get_chunk_info(chunk_hash)?.chunk_header)
+    }
+
+    pub fn chunk_endorsements(
+        &self,
+        chunk_hash: &ChunkHash,
+    ) -> Result<ChunkEndorsementSignatures, Error> {
+        Ok(self.get_chunk_info(chunk_hash)?.signatures.clone().unwrap_or_default())
+    }
+
+    pub fn chunk_producer_and_received_time(
+        &self,
+        chunk_hash: &ChunkHash,
+    ) -> Result<(AccountId, DateTime<Utc>), Error> {
+        let chunk_info = self.get_chunk_info(chunk_hash)?;
+        Ok((chunk_info.chunk_producer.clone(), chunk_info.received_time))
+    }
+}
+
+fn filter_banned_chunk_producers(
+    banned_chunk_producers: &LruCache<(EpochId, AccountId), ()>,
+    epoch_id: &EpochId,
+    chunk_info: &ChunkInfo,
+) -> bool {
+    let banned =
+        banned_chunk_producers.contains(&(epoch_id.clone(), chunk_info.chunk_producer.clone()));
+    if banned {
+        tracing::warn!(
+            target: "client",
+            chunk_hash = ?chunk_info.chunk_header.chunk_hash(),
+            chunk_producer = ?chunk_info.chunk_producer,
+            "Not including chunk from a banned validator");
+        metrics::CHUNK_DROPPED_BECAUSE_OF_BANNED_CHUNK_PRODUCER.inc();
+    }
+    !banned
+}
+
+fn filter_insufficient_chunk_endorsements(
```

Comment:
I have concerns about `filter_` functions here:
1) I think usually when we say `filter` then it is about selecting subset of some set, but here it is about saying yes or no, so it is better to call it `has_sufficient_chunk_endorsements`, etc.
2) I would avoid it to have side effect of pulling endorsements from `ChunkValidator`. First, it took a while to understand where it happens; second, I think readers usually don't assume that functions with boolean result also mutate its internal state.

What about updating endorsements explicitly, before `num_chunk_headers_ready_for_inclusion` on block production? Then we don't have to mutate anything here.

Reply:
Renamed function, removed `mut` from them. Added explicit step `prepare_chunk_headers_ready_for_inclusion`